File: tests/fixtures/vulnerable/build-then-run.go
// THE TWO-LINE FORM, IN GO.
//
// See build-then-run.py. Go only reports exec.Command when the call itself
// names a shell - exec.Command("ls", name) is an argument vector and genuinely
// safe - and that rule is unchanged. What changes is that the command handed
// to `sh -c` is now followed back to where it was built.
package fixtures

import (
	"fmt"
	"os/exec"
)

func concatThenRun(name string) {
	c := "ls " + name
	// EXPECT-SIGNATURE command-injection
	exec.Command("sh", "-c", c).Run()
}

func sprintfThenRun(host string) {
	c := fmt.Sprintf("ping -c 1 %s", host)
	// EXPECT-SIGNATURE command-injection
	exec.Command("bash", "-c", c).Run()
}

// ---------------------------------------------------------------- FENCES

func fixed() {
	c := "uptime"
	exec.Command("sh", "-c", c).Run()
}

// No shell: an argument vector, so `name` is one argument however it is spelled.
func argumentVector(name string) {
	exec.Command("ls", "-l", name).Run()
}

File: tests/fixtures/vulnerable/BuildThenRun.java
/*
 * THE TWO-LINE FORM, IN JAVA.
 *
 * See build-then-run.py. The Java coverage note already admitted the signature
 * pass only saw commands "built at the call site"; the single-statement build
 * one line earlier is now followed. StringBuilder chains across several
 * statements are still not, and the note still says so.
 */
public class BuildThenRun {

    void concatThenRun(String name) throws Exception {
        String cmd = "ls -l " + name;
        // EXPECT-SIGNATURE command-injection
        Runtime.getRuntime().exec(cmd);
    }

    void fixed() throws Exception {
        String cmd = "uptime";
        Runtime.getRuntime().exec(cmd);
    }
}
